### PR TITLE
white noise relaxed clock

### PR DESCRIPTION
**Pull request by zhangchicool**
Brought back the old 'IGR' relaxed clock model and renamed it as 'WN' (white noise). The name 'IGR' now refers to a new model where the gamma variance is just 'v' (does not depend on branch length).

### CODE REVIEW

File: src/command.c [hunk 1]
```diff
@@ -12705,9 +12742,11 @@ else if (!strcmp(helpTkn, "Set"))
         MrBayesPrint ("      Cppmultdev      -- Standard dev. of CPP rate multipliers (log scale)       \n"); 
         MrBayesPrint ("      Cppevents       -- CPP events                                              \n"); 
         MrBayesPrint ("      TK02var         -- Variance increase in TK02 relaxed clock model           \n"); 
+        MrBayesPrint ("      WNvar           -- Variance increase in WN relaxed clock model             \n");
         MrBayesPrint ("      Igrvar          -- Variance increase in IGR relaxed clock model            \n");
```
Comment (ronquist) [inline src/command.c]: Correct to "Variance in IGR relaxed clock model". There is no variance increase with branch length in this model.

File: src/command.c [hunk 2]
```diff
@@ -12705,9 +12742,11 @@ else if (!strcmp(helpTkn, "Set"))
...
         MrBayesPrint ("      Ilnvar          -- Variance increase in ILN relaxed clock model            \n");
```
Comment (ronquist) [inline src/command.c]: Same as above, correct to "Variance in ILN relaxed clock model".

File: src/command.c [hunk 3]
```diff
@@ -12753,9 +12792,11 @@ else if (!strcmp(helpTkn, "Set"))
         MrBayesPrint ("      Cppmultdev      -- Standard dev. of CPP rate multipliers (log scale)       \n");
         MrBayesPrint ("      Cppevents       -- CPP events                                              \n");
         MrBayesPrint ("      TK02var         -- Variance increase in TK02 relaxed clock model           \n");
+        MrBayesPrint ("      WNvar           -- Variance increase in WN relaxed clock model             \n");
         MrBayesPrint ("      Igrvar          -- Variance increase in IGR relaxed clock model            \n");
```
Comment (ronquist) [inline src/command.c]: Correct to "Variance in IGR relaxed clock model"

File: src/command.c [hunk 4]
```diff
@@ -12753,9 +12792,11 @@ else if (!strcmp(helpTkn, "Set"))
...
         MrBayesPrint ("      Ilnvar          -- Variance increase in ILN relaxed clock model            \n");
```
Comment (ronquist) [inline src/command.c]: Correct to "Variance of ILN relaxed clock model"

File: src/utils.c [hunk 5]
```diff
@@ -4906,30 +4906,17 @@ int IsTreeConsistent (Param *param, int chain, int state)
...
+            {    // *GetParamIntVals(subParm, chain, state) == RCL_ILN or RCL_IGR
             for (j=0; j<tree->nNodes-2; j++)
                 {
                 p = tree->allDownPass[j];
                 b = GetParamSubVals(subParm, chain, state)[p->index];
                 r = GetParamVals(subParm, chain, state)[p->index];
                 if (fabs(p->length * r - b) > 0.000001)
                     {
-                    printf ("%s   Igr relaxed clock mismatch in branch %d\n", spacer, p->index);
+                    printf ("%s   Independent relaxed clock mismatch in branch %d\n", spacer, p->index);
```
Comment (ronquist) [inline src/utils.c]: Correct to "Independent/uncorrelated relaxed clock mismatch..."

File: src/command.c [hunk 6]
```diff
@@ -11196,23 +11196,25 @@ int GetUserHelp (char *helpTkn)
...
-        MrBayesPrint ("                    'tk02' invokes the Brownian Motion model described by Thorne \n");
-        MrBayesPrint ("                    and Kishino (2002). [autocorrelated lognormal distributions] \n");
-        MrBayesPrint ("                    'igr' invokes the Independent Gamma Rate (IGR) model where   \n");
-        MrBayesPrint ("                    each branch has an independent rate drawn from a gamma       \n");
-        MrBayesPrint ("                    distribution (LePage et al., 2007).                          \n");
+        MrBayesPrint ("                    'tk02' invokes the geometric Brownian Motion model described \n");
+        MrBayesPrint ("                    by Thorne and Kishino (2002). [autocorrelated lognormal]     \n");
+        MrBayesPrint ("                    'wn' invokes the white noise model (LePage et al., 2007)     \n");
+        MrBayesPrint ("                    where each branch has an independent rate drawn from a gamma \n");
+        MrBayesPrint ("                    distribution with variance propotional to the branch length. \n");
```
Comment (ronquist) [inline src/command.c]: Correct "propotional" to "proportional"

File: src/command.c [hunk 7]
```diff
@@ -11196,23 +11196,25 @@ int GetUserHelp (char *helpTkn)
...
+        MrBayesPrint ("                    'igr' invokes the Independent Gamma Rate model. The differ-  \n");
+        MrBayesPrint ("                    ence from 'wn' is that these gamma distributions are i.i.d.  \n");
+        MrBayesPrint ("                    thus the variances do not depend on the branch lengths.      \n");
```
Comment (ronquist) [inline src/command.c]: Correct to "Thus, the variances..."

File: src/command.c [hunk 8]
```diff
@@ -11196,23 +11196,25 @@ int GetUserHelp (char *helpTkn)
...
+        MrBayesPrint ("                    'iln' invokes the Independent Lognormal model. It is similar \n");
+        MrBayesPrint ("                    with 'igr' but the rates are i.i.d. lognormal distributions.\n");
```
Comment (ronquist) [inline src/command.c]: Correct to "to 'igr' but the..."

File: src/command.c [hunk 9]
```diff
@@ -11196,23 +11196,25 @@ int GetUserHelp (char *helpTkn)
...
         MrBayesPrint ("                    Each of the relaxed clock models has additional parameters   \n");
         MrBayesPrint ("                    with priors. For the CPP model, it is 'cppratepr' and        \n");
         MrBayesPrint ("                    'cppmultdevpr'; for the TK02 model, it is 'tk02varpr'; for   \n");
         MrBayesPrint ("                    the IGR  model, it is 'igrvarpr'.                            \n");
```
Comment (ronquist) [inline src/command.c]: Replace this line with "the WN model, it is 'wnvarpr'; for the IGR model, it is 'igrvarpr'; for the ILN model, it is 'ilnvarpr'.

File: src/model.c [hunk 10]
```diff
@@ -24200,10 +24534,12 @@ int ShowParameters (int showStartVals, int showMoves, int showAllAvailable)
                 MrBayesPrint ("%s                         The clock rate varies according to a CPP model\n", spacer);
             else if (!strcmp(mp->clockVarPr,"TK02"))
                 MrBayesPrint ("%s                         The clock rate varies according to a autocorrelated lognromal model\n", spacer);
```
Comment (ronquist) [inline src/model.c]: Correct "lognromal" to "lognormal"

File: src/model.c [hunk 11]
```diff
@@ -24200,10 +24534,12 @@ int ShowParameters (int showStartVals, int showMoves, int showAllAvailable)
...
+            else if (!strcmp(mp->clockVarPr,"WN"))
+                MrBayesPrint ("%s                         The clock rate varies according to a white noise model\n", spacer);
+            else if (!strcmp(mp->clockVarPr,"Igr"))
+                MrBayesPrint ("%s                         The clock rate varies according to an independent gamma model\n", spacer);
             else if (!strcmp(mp->clockVarPr,"Iln"))
                 MrBayesPrint ("%s                         The clock rate varies according to an independent lognromal model\n", spacer);
```
Comment (ronquist) [inline src/model.c]: Correct "lognromal" to "lognormal"

File: src/proposal.c [hunk 12]
```diff
@@ -2226,6 +2322,26 @@ int Move_ExtFossilSPRClock (Param *param, int chain, RandLong *seed, MrBFlt *lnP
             brlens[v->index] = v->length * (rate[v->index] + rate[v->anc->index]) / 2.0;
             brlens[u->index] = u->length * (rate[u->index] + rate[u->anc->index]) / 2.0;
         }   /* end tk02 branch rate parameter */
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+        {
+            /* adjust prior ratio */
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c]: These four lines should probably be replaced by just one line, line 2329. This is the only line that will be executed anyway.

File: src/proposal.c [hunk 13]
```diff
@@ -3444,6 +3560,25 @@ int Move_ExtSPRClock (Param *param, int chain, RandLong *seed, MrBFlt *lnPriorRa
             brlens = GetParamSubVals (subParm, chain, state[chain]);
             brlens[a->index] = (rate[a->index] + rate[b->index]) / 2.0 * (a->length + u->length);
             }   /* end tk02 branch rate parameter */
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c:3269]: Looks like copy and paste without correction, see above. All four lines should probably be replaced by line 3566.

File: src/proposal.c [hunk 14]
```diff
@@ -3693,6 +3828,26 @@ int Move_ExtSPRClock (Param *param, int chain, RandLong *seed, MrBFlt *lnPriorRa
             brlens[v->index] = v->length * (rate[v->index] + rate[v->anc->index]) / 2.0;
             brlens[u->index] = u->length * (rate[u->index] + rate[u->anc->index]) / 2.0;
             }   /* end tk02 branch rate parameter */
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            /* adjust prior ratio */
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c:3801]: Same comment as above. Is this from the mixed clock model?

File: src/proposal.c [hunk 15]
```diff
@@ -4637,6 +4792,24 @@ int Move_ExtSSClock (Param *param, int chain, RandLong *seed, MrBFlt *lnPriorRat
             brlens[a->index] = a->length * (rate[a->index] + rate[a->anc->index])/2.0;
             brlens[c->index] = c->length * (rate[c->index] + rate[c->anc->index])/2.0;
             }
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            /* get relevant parameters */
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c:4802]: Replace these four lines with one line, see comments above.

File: src/proposal.c [hunk 16]
```diff
@@ -7913,6 +8086,22 @@ int Move_NNIClock (Param *param, int chain, RandLong *seed, MrBFlt *lnPriorRatio
             brlens[a->index] = a->length * (rate[a->index] + rate[a->anc->index])/2.0;
             brlens[c->index] = c->length * (rate[c->index] + rate[c->anc->index])/2.0;
             }
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c:8095]: Replace with one line, see above.

File: src/proposal.c [hunk 17]
```diff
@@ -8432,6 +8621,38 @@ int Move_NodeSliderClock (Param *param, int chain, RandLong *seed, MrBFlt *lnPri
                 brlens[p->index] = p->length * (rate[p->index]+rate[p->anc->index])/2.0;
                 }
             }
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c:8630]: Replace with one line, see above.

File: src/proposal.c [hunk 18]
```diff
@@ -9800,6 +10021,25 @@ int Move_ParsFossilSPRClock (Param *param, int chain, RandLong *seed, MrBFlt *ln
             brlens = GetParamSubVals (subParm, chain, state[chain]);
             brlens[a->index] = (rate[a->index] + rate[b->index]) / 2.0 * (a->length + u->length);
         }   /* end tk02 branch rate parameter */
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+        {
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c]: Replace with one line, see above.

File: src/proposal.c [hunk 19]
```diff
@@ -10169,6 +10409,27 @@ int Move_ParsFossilSPRClock (Param *param, int chain, RandLong *seed, MrBFlt *ln
             brlens[v->index] = v->length * (rate[v->index] + rate[v->anc->index]) / 2.0;
             brlens[u->index] = u->length * (rate[u->index] + rate[u->anc->index]) / 2.0;
         }   /* end tk02 branch rate parameter */
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+        {
+            /* adjust prior ratio */
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c]: Replace with one line, see above.

File: src/proposal.c [hunk 20]
```diff
@@ -12208,6 +12469,25 @@ int Move_ParsSPRClock (Param *param, int chain, RandLong *seed, MrBFlt *lnPriorR
             brlens = GetParamSubVals (subParm, chain, state[chain]);
             brlens[a->index] = (rate[a->index] + rate[b->index]) / 2.0 * (a->length + u->length);
             }   /* end tk02 branch rate parameter */
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c]: Replace with one line, see above.

File: src/proposal.c [hunk 21]
```diff
@@ -12577,6 +12857,26 @@ int Move_ParsSPRClock (Param *param, int chain, RandLong *seed, MrBFlt *lnPriorR
             brlens[v->index] = v->length * (rate[v->index] + rate[v->anc->index]) / 2.0;
             brlens[u->index] = u->length * (rate[u->index] + rate[u->anc->index]) / 2.0;
             }   /* end tk02 branch rate parameter */
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            /* adjust prior ratio */
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c:12842]: Replace with one line.

File: src/proposal.c [hunk 22]
```diff
@@ -16407,6 +16707,33 @@ int Move_TreeStretch (Param *param, int chain, RandLong *seed, MrBFlt *lnPriorRa
                     }
                 }
             }
+        else if ( subParm->paramType == P_WNBRANCHRATES)
+            {
+            if (subParm->paramType == P_WNBRANCHRATES)
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].wnvar, chain, state[chain]);
+            else
+                var = *GetParamVals (modelSettings[subParm->relParts[0]].mixedvar, chain, state[chain]);
```
Comment (ronquist) [inline src/proposal.c:16715]: Replace with one line, see above.

**Review by ronquist — CHANGES_REQUESTED** [review]

**Comment by ronquist** [comment]
Chi, great if you can review the requested changes and then change the code accordingly. Can you please help us also add a text snippet describing the changes you introduced in the document outlining what is new in the next release? Andreas can help you with instructions if needed.

File: src/proposal.c (hunk 12 above)
Comment (zhangchicool) [reply src/proposal.c]: I would just leave it as is. Maybe it will be useful when more relaxed clock models are included in the reversible jump.

**Comment by zhangchicool** [comment]
I have made the changes accordingly, please re-review, will do the release note later.

**Review by ronquist — APPROVED** [review]
Looks great now!

**Pull request merged**